File: MAP/Lab 2/ToyInterpreter/src/repository/Repository.java
package repository;

import exceptions.AdtExceptions;
import model.adts.MyDictionary;
import model.adts.MyList;
import model.adts.MyStack;
import model.state.FileTableData;
import model.state.PrgState;
import model.statements.IStmt;

import java.io.*;
import java.util.Scanner;

public class Repository implements IRepository, IRepositoryLog{
    private MyList<PrgState> programs;
    private String logFilePath;

    public Repository() {
        this.programs = new MyList<PrgState>();
        initRepoLogFilePath();
    }

    public Repository(String path) {
        this.programs = new MyList<PrgState>();
        logFilePath = path;
    }

    private void initRepoLogFilePath() {
        System.out.print("Repository log file path: ");
        Scanner sc = new Scanner(System.in);
        logFilePath = sc.nextLine();
    }

    public MyList<PrgState> getPrograms() {
        return programs;
    }

    @Override
    public void setPrograms(MyList<PrgState> programs) {
        this.programs = programs;
    }

    @Override
    public PrgState addProgram(IStmt stmt) throws Exception {
        if(programs.size() != 0)
            throw new Exception("Only one program can be added");

        PrgState newProg = new PrgState(
                new MyStack<>(),
                new MyDictionary<String, Integer>(),
                new MyList<Integer>(),
                stmt,
                new MyDictionary<Integer, FileTableData>(),
                new MyDictionary<Integer, Integer>(),
                1
        );

        programs.push_back(newProg);

        return newProg;
    }

    @Override
    public void logPrgStateExec(PrgState state) throws IOException, AdtExceptions {
        PrintWriter logFile = new PrintWriter(new BufferedWriter(new FileWriter(logFilePath, true)));
        logFile.append("\n=== Prog Eval (ID "+ state.getId() +") ===\n");

        logFile.append("ExeStack:\n");
        logFile.append(state.getExeStack().toString() + "\n");
        logFile.append("SymTable:\n");
        logFile.append(state.getSymTable().toString() + "\n");
        logFile.append("Out:\n");
        logFile.append(state.getOut().toString() + "\n");
        logFile.append("FileTable:\n");
        logFile.append(state.getFileTable().toString() + "\n");
        logFile.append("Heap: \n");
        logFile.append(state.getHeap().toString() + "\n");

        logFile.flush();
    }
}

File: MAP/Lab10/ConsoleApp1/ConsoleApp1/Repository/Repository.cs
using ConsoleApp1.Model.state;
using System;
using System.Collections.Generic;
using System.Text;
using ConsoleApp1.Model.statements;

namespace ToyInterpreter.Repository
{
    class Repository
    {
        public List<PrgState> programs;
        private string logFilePath;

        public Repository()
        {
            this.programs = new List<PrgState>();
            readLogFilePath();
        }

        public PrgState getCrtPrg()
        {
            if (programs.Count == 0)
                throw new Exception("There must be a program inserted");
            
            return this.programs[0];
        }

        public void addProgram(PrgState prg)
        {
            if (programs.Count == 0)
                this.programs.Add(prg);
            else
                throw new Exception("Only one program must exist");
        }

        public void readLogFilePath()
        {
            Console.WriteLine("Log file path: ");
            this.logFilePath = Console.ReadLine();
        }

        public void logPrgStateExec(PrgState state) {
            string logFileContent = "";

            logFileContent += state.ToString();

            System.IO.File.AppendAllText(this.logFilePath, logFileContent);
        }
}
}
